credoai/lens/lens_validation.py

```diff
@@ -0,0 +1,172 @@
+import pandas as pd
+
+import numpy as np
+
+import tensorflow as tf
+
+import inspect
+
+from credoai.utils.common import ValidationError
+from credoai.utils import global_logger
+
+
+###############################################
+# Checking artifact interactions (model + data)
+###############################################
+
+
+def check_model_data_consistency(model, data):
+    """
+    This validation function serves to check the compatibility of a model and dataset provided to Lens.
+    For each outputting function (e.g., `predict`) supported by Lens, this validator checks
+    whether the model supports that function. If so, the validator applies the outputting function to
+    a small sample of the supplied dataset. The validator ensures the outputting function does not fail
+    and, depending on the nature of the outputting function, performs light checking to verify the outputs
+    (e.g. predictions) match the expected form, possibly including: data type and output shape.
+
+    Parameters
+    ----------
+    model : artifacts.Model or a subtype of artifacts.Model
+        A trained machine learning model wrapped as a Lens Model object
+    data : artifacts.Data or a subtype of artifacts.Data
+        The dataset that will be assessed by Lens evaluators, wrapped as a Lens Data object
+    """
+    # check predict
+    # Keras always outputs numpy types (not Tensor or something else)
+    if "predict" in model.__dict__.keys():
+        try:
+            mini_pred, batch_size = check_prediction_model_output(model.predict, data)
+            if not mini_pred.size:
+                # results for all presently supported models are ndarray results
+                raise Exception("Empty return results from predict function.")
+            if isinstance(data.y, np.ndarray) and (
+                mini_pred.shape != data.y[:batch_size].shape
+            ):
+                raise Exception("Predictions have mismatched shape from provided y")
+            if isinstance(data.y, pd.Series) and (
+                mini_pred.shape != data.y.head(batch_size).shape
+            ):
+                raise Exception("Predictions have mismatched shape from provided y")
+        except Exception as e:
+            raise ValidationError(
+                "Lens.model predictions do not match expected form implied by provided labels y.",
```

Comment:
Isn't this error too specific? If the predict function flat out fails, don't we want to say that and reproduce the error? I'm worried we are catching an exception that we want to raise.

Reply:
The call to the predict function happens in the helper function on line 38. There's no specific error message in there so if the predict function fails, then we'll catch and raise the message on line 52 + the original exception (note the addition of the exception with `, e)` on line 53/54.

The checks on lines 39-49 kick in only if `predict` itself works but there's somehow a mismatch between `y_pred` and the supplied `y_true`.